frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionResult.java

```diff
@@ -18,30 +18,65 @@
 package com.microsoft.frameworklauncher.applicationmaster;
 
 import com.microsoft.frameworklauncher.common.exts.CommonExts;
+import com.microsoft.frameworklauncher.common.model.ValueRange;
+import com.microsoft.frameworklauncher.common.model.ResourceDescriptor;
+import com.microsoft.frameworklauncher.common.utils.ValueRangeUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.List;
 
 public class SelectionResult {
-  private String nodeHost;
-  private Long gpuAttribute;
 
-  public String getNodeHost() {
-    return nodeHost;
+  private Map<String, Long> nodes = new HashMap<String, Long>();
+  private List<ValueRange> overlapPorts = new ArrayList<ValueRange>();
+  private ResourceDescriptor optimizedResource = ResourceDescriptor.newInstance(0, 0, 0, (long) 0);
+
+  public void addSelection(String hostName, Long gpuAttribute, List<ValueRange> portList) {
+    if (nodes.isEmpty()) {
+      nodes.put(hostName, gpuAttribute);
+      overlapPorts = ValueRangeUtils.coalesceRangeList(portList);
+      return;
+    }
+    if (nodes.containsKey(hostName)) {
+      nodes.remove(hostName);
+    }
+    nodes.put(hostName, gpuAttribute);
+    overlapPorts = ValueRangeUtils.intersectRangeList(overlapPorts, portList);
+  }
+
+  public List<ValueRange> getOverlapPorts() {
+    return overlapPorts;
+  }
+
+  public List<String> getNodeHosts() {
+    List<String> hostList = new ArrayList<String>();
+    for (String hostName : nodes.keySet()) {
+      hostList.add(hostName);
+
+    }
+    return hostList;
   }
 
-  public void setNodeHost(String nodeHost) {
-    this.nodeHost = nodeHost;
+  public Long getGpuAttribute(String hostName) {
+    return nodes.get(hostName);
   }
 
-  public Long getGpuAttribute() {
-    return gpuAttribute;
+  public void setOptimizedResource(ResourceDescriptor optimizedResource) {
+    this.optimizedResource = optimizedResource;
   }
 
-  public void setGpuAttribute(Long gpuAttribute) {
-    this.gpuAttribute = gpuAttribute;
+  public ResourceDescriptor getOptimizedResource() {
+    return optimizedResource;
   }
 
   @Override
   public String toString() {
-    return String.format("[NodeHost: [%s]", getNodeHost()) + " " +
-        String.format("GpuAttribute: [%s]]", CommonExts.toStringWithBits(getGpuAttribute()));
+    String output = "SelectionResult:";
+    for (Map.Entry<String, Long> entry : nodes.entrySet()) {
+      output += String.format(" [Host: %s GpuAttribute: %s]", entry.getKey(), CommonExts.toStringWithBits(entry.getValue()));
+    }
```

Comment:
Also include overlapPorts and optimizedResource?

Reply:
the string is too long to included here. 

---
In reply to: [174692468](https://github.com/Microsoft/pai/pull/318#discussion_r174692468) [](ancestors = 174692468)

Reply:
optimizedResource should be short?

---
In reply to: [175341843](https://github.com/Microsoft/pai/pull/318#discussion_r175341843) [](ancestors = 175341843,174692468)

Reply:
OptimizedResource

---
In reply to: [175380121](https://github.com/Microsoft/pai/pull/318#discussion_r175380121) [](ancestors = 175380121,175341843,174692468)